contentcuration/contentcuration/frontend/shared/vuex/channel/mutations.js

```diff
@@ -51,9 +51,13 @@ export function UPDATE_CHANNEL(state, { id, content_defaults = {}, ...payload }
     });
   }
 }
+
 export function UPDATE_CHANNEL_FROM_INDEXEDDB(state, { id, ...mods }) {
   if (id && state.channelsMap[id]) {
-    applyMods(state.channelsMap[id], mods);
+    state.channelsMap = {
+      ...state.channelsMap,
+      [id]: applyMods(state.channelsMap[id], mods),
+    };
```

Comment:
I realize Vue.js is the major factor in causing these changes. This has bad code smell to me though. Reassigning a variable containing an object, to itself, even though the object was already modified really seems like an anti-pattern. We only modified a subset of a nested object, yet we're completely reassigning the variable. It also hides the implicit reasons for doing this since it shouldn't be needed when thinking in terms of JavaScript. (side note: this is one of my gripes with Vue.js and the main reason I dislike it-- it requires the programmer to write code the Vue.js way, which goes against the grain of standard JS with a component API that is wholly unfamiliar compared to common OOP structures, and functionality that causes contradictions like this in behavior inherent to JS)

Our override of `Dexie.setByKeyPath` returns the object passed in, but the original implementation doesn't. If `applyMods` modifies the object passed in, it would make the function's API clearer if it didn't also return it. Either it should return a new object, or modify the passed in object and return nothing. If `applyMods` did return a new object, the following bit of code would solve the issue at hand:
```js
Vue.set(state.channelsMap, id, applyMods(state.channelsMap[id], mods))
```
or without refactoring (and how I tested this):
```js
Vue.set(state.channelsMap, id, {...applyMods(state.channelsMap[id], mods)})
```
I think the above makes it clearer that we're invoking behavior of Vue, that shouldn't otherwise be required in JS when modifying objects, thus lessening the smell of this code. So this object spread pattern, with it's implicit reasons and compared with an alternative, does seem like an anti-pattern to me.

Reply:
Both `Vue.set` and reassigning with object spread are the two ways explicitly recommended by Vuex in their documentation, which is why this pattern is used here and in lots of other places: https://v3.vuex.vuejs.org/guide/mutations.html#mutations-follow-vue-s-reactivity-rules

My only concern is that it's not completely clear to me how this operates in Vuex when setting objects as the values - as far as I can tell from the implementation of `Vue.set` (https://github.com/vuejs/vue/blob/2.6/src/core/observer/index.js#L201) and how it is used in Vuex (https://github.com/vuejs/vuex/blob/3.x/src/store.js#L355) there doesn't seem to be any flagging of `deep` for the setting/watching - however, I don't know if this will be different between `Vue.set` and object spread either.

I'm going to add some tests specifically for the contentnode mutations to ensure that our nested updates in Vuex are being properly watched to remove any doubt here.

Reply:
Yeah I think since the documentation says
> When adding new properties to an Object, you should either...

the notable difference here is that the `[id]` key isn't being added, since it should already exist. So on your point regarding the `deep` setting, that seems like it would be an issue with what you've done here. Also, I think my main point was the API of `applyMods` and that creating a new object (like the spread would do on the channel object) should be able handle this with a clearer API overall.

> explicitly recommended by Vuex in their documentation, which is why this pattern is used here and in lots of other places

I get that, but I'm saying the spread pattern has an alternative, also listed in the documentation, with higher precedence than the spread example. I don't believe that just because the docs say it that we should use the pattern! I think we should strive for self-documenting code, and using `Vue.set` makes the purpose of the code clearer.

Also, I suspect there must be a difference between the two options with regards to how the garbage collector reacts, since we'd be creating new objects at the `channelsMap` level instead of just the affected channel itself.